brion/detail/serializableMorphology.ipp

```diff
@@ -0,0 +1,164 @@
+    memcpy(dst, src->data(), sizeof(T) * src->size());
+    dst += sizeof(T) * src->size();
+}
+
+template <typename T>
+bool _deserializeArray(std::shared_ptr<std::vector<T>>& dst,
+                       const uint8_t*& src, const uint8_t* end)
+{
+    if (src + sizeof(uint64_t) > end)
+        return false;
+    const uint64_t arraySize = *reinterpret_cast<const uint64_t*>(src);
+    src += sizeof(uint64_t);
+
+    if (src + sizeof(T) * arraySize > end)
+        return false;
+    const T* dstPtr = reinterpret_cast<const T*>(src);
+    dst.reset(new std::vector<T>(dstPtr, dstPtr + arraySize));
+    src += sizeof(T) * arraySize;
+    return true;
+}
+}
+
+inline SerializableMorphology::SerializableMorphology(const void* data,
+                                                      const size_t size)
+{
+    if (!_fromBinary(data, size))
+        LBTHROW(std::runtime_error(
+            "Failed to construct morphology from binary data"));
+}
+
+inline SerializableMorphology::SerializableMorphology(brion::Morphology& m)
+    : _points(m.readPoints())
+    , _sections(m.readSections())
+    , _types(m.readSectionTypes())
+    , _apicals(m.readApicals())
+    , _perimeters(m.readPerimeters())
+{
+}
+
+inline SerializableMorphology::~SerializableMorphology()
+{
+}
+
+void inline SerializableMorphology::setPoints(const Vector4fs& points)
+{
+    _points.reset(new Vector4fs(points));
+}
+
+void inline SerializableMorphology::setSections(const Vector2is& sections)
+{
+    _sections.reset(new Vector2is(sections));
+}
+
+void inline SerializableMorphology::setSectionTypes(const SectionTypes& types)
+{
+    _types.reset(new SectionTypes(types));
+}
+
+void inline SerializableMorphology::setApicals(const Vector2is& apicals)
+{
+    _apicals.reset(new Vector2is(apicals));
+}
+
+void inline SerializableMorphology::setPerimeters(const floats& perimeters)
+{
+    _perimeters.reset(new floats(perimeters));
+}
+
+servus::Serializable::Data inline SerializableMorphology::_toBinary() const
+{
+    servus::Serializable::Data data;
+
+    data.size = sizeof(MorphologyVersion) + sizeof(CellFamily) +
+                sizeof(uint64_t) + sizeof(brion::Vector4f) * _points->size() +
+                sizeof(uint64_t) + sizeof(brion::Vector2i) * _sections->size() +
+                sizeof(uint64_t) + sizeof(uint32_t) * _types->size() +
+                sizeof(uint64_t) + sizeof(brion::Vector2i) * _apicals->size() +
+                sizeof(uint64_t) + sizeof(float) * _perimeters->size();
+
+    uint8_t* ptr = new uint8_t[data.size];
+    data.ptr.reset(ptr, std::default_delete<uint8_t[]>());
+
+    *reinterpret_cast<MorphologyVersion*>(ptr) = _version;
+    ptr += sizeof(MorphologyVersion);
+
+    *reinterpret_cast<CellFamily*>(ptr) = _family;
+    ptr += sizeof(CellFamily);
+
+    _serializeArray(ptr, _points);
+    _serializeArray(ptr, _sections);
+    _serializeArray(ptr, _types);
+    _serializeArray(ptr, _apicals);
+    _serializeArray(ptr, _perimeters);
```

Comment:
OK, but this is not a plugin detail, but a data model detail. Client code using Brion also needs to take it into account.